### PR TITLE
Typescripting the 3D-Signalstk

### PR DESCRIPTION
**Pull request by conrad465**
- Porting all functionality to TS and defining types and interfaces for maintainability.
- fixing funky importing of webrtc-signal-http after porting that to TS

TODO - one test is failing due to an error unrelated to the test case FIXED

### CODE REVIEW

File: dist/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env node
```
Comment (anderm) [inline dist/index.d.ts]: Build artifacts should **not** be checked in to source control. This includes the dist directory

File: lib/index.ts [hunk 2]
```diff
@@ -1,147 +1,147 @@
...
-    // return a configured app, ready to be used
-    return app
-}
+import * as express from "express";
+import * as expressBunyan from "express-bunyan-logger";
+import * as passport from "passport";
+import { ITokenPayload, BearerStrategy, VerifyCallback, IBearerStrategyOption } from "passport-azure-ad";
+import { signalRouterCreator, IPeerRequest, PeerList } from "webrtc-signal-http-ts";
+import { ISignalerOpts, IPeerBearerStrategyOpts, IExpressApp } from "./modules";
+const heartbeatRouterCreator = require("webrtc-signal-http-heartbeat");
```
Comment (anderm) [inline lib/index.ts]: We should be consistent, either use the `require` pattern or `import`. Since we moved to ts, we should change all to `import`

File: lib/index.ts [hunk 3]
```diff
@@ -1,147 +1,147 @@
...
+
+const optIsFalsey = (opt: string | boolean) => {
+    return !opt ||
+        opt === 'false' ||
+        (typeof(opt) === 'string' && opt.toLowerCase() === 'false') 
+}
+
+module.exports = (opts: ISignalerOpts) => {
+    const app = express() as IExpressApp; 
+    let peerList = new PeerList() as PeerList;
```
Comment (anderm) [inline lib/index.ts]: no need to double cast

File: lib/index.ts [hunk 4]
```diff
@@ -1,147 +1,147 @@
...
+const Publisher = require("webrtc-signal-http-publisher");
+
+const optIsFalsey = (opt: string | boolean) => {
+    return !opt ||
+        opt === 'false' ||
+        (typeof(opt) === 'string' && opt.toLowerCase() === 'false') 
+}
+
+module.exports = (opts: ISignalerOpts) => {
+    const app = express() as IExpressApp;
```
Comment (anderm) [inline lib/index.ts]: these lines end with `;` but the rest of the file does not. We should be consistent

File: lib/index.ts [hunk 5]
```diff
@@ -1,147 +1,147 @@
...
+        })
+
+        // use the recognition peer list instead
+        peerList = recognitionRouter.peerList
+
+        app.use(recognitionRouter)
+    }
+
+    // configure the webrtc-signal-http router
+    const signalRouter = signalRouterCreator({...opts,
```
Comment (anderm) [inline lib/index.ts]: Move `...opts` to a new line

File: lib/index.ts [hunk 6]
```diff
@@ -1,147 +1,147 @@
...
+import { ISignalerOpts, IPeerBearerStrategyOpts, IExpressApp } from "./modules";
+const heartbeatRouterCreator = require("webrtc-signal-http-heartbeat");
+const capacityRouterCreator = require("webrtc-signal-http-capacity");
+const recognitionRouterCreator = require("webrtc-signal-http-peer-identification");
+const Publisher = require("webrtc-signal-http-publisher");
+
+const optIsFalsey = (opt: string | boolean) => {
+    return !opt ||
+        opt === 'false' ||
+        (typeof(opt) === 'string' && opt.toLowerCase() === 'false')
```
Comment (anderm) [inline lib/index.ts]: nitpick: `typeof` should have a space after

File: lib/index.ts [hunk 7]
```diff
@@ -1,147 +1,147 @@
...
+    // configure heartbeat if needed
+    if (!optIsFalsey(opts.heartbeatEnabled)) {
+        const heartbeatRouter = heartbeatRouterCreator({
+            peerList: peerList,
+            timeoutPeriod: opts.heartbeatMs || 30 * 1000,
+            gcInterval: opts.heartbeatGcMs || 15 * 1000
+        })
+
+        // use the heartbeat peerList instead
+        peerList = heartbeatRouter.peerList as PeerList
```
Comment (anderm) [inline lib/index.ts]: no need to cast to PeerList

File: lib/modules.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { IBearerStrategyOption } from "passport-azure-ad";
+import  { Express } from "express";
+import { PeerList } from "webrtc-signal-http-ts";
+import { EventEmitter } from "events";
+const Publisher = require("webrtc-signal-http-publisher");
```
Comment (anderm) [inline lib/modules.ts]: Same here, switch to `import`

File: test/basic.ts [hunk 9]
```diff
@@ -1,201 +1,201 @@
...
-                .get('/sign_in?peer_name=testName')
-                .set('x-forwarded-for', '::2')
-                .expect(200, () => {
-                    assert.equal(app.peerList.getPeer(1).ip, '::2')
-                    done()
-                })
-        })
-    })
-})
+const assert = require('assert')
```
Comment (anderm) [inline test/basic.ts]: Switch to `import`

File: index.ts [hunk 10]
```diff
@@ -1,23 +1,23 @@
...
-    authTenantId: process.env.WEBRTC_AUTH_TENANT_ID,
-    capacityEnabled: process.env.WEBRTC_CAPACITY_ENABLED,
-    recognitionEnabled: process.env.WEBRTC_RECOGNITION_ENABLED,
-    publishState: process.env.WEBRTC_PUBLISH_STATE,
-    trustProxy: process.env.WEBRTC_TRUST_PROXY,
-    enableCors: process.env.WEBRTC_CORS
-})
-
+#!/usr/bin/env node
+const appCreator = require('./lib')
```
Comment (anderm) [inline index.ts]: Switch to import

File: index.ts [hunk 11]
```diff
@@ -1,23 +1,23 @@
...
+
+const app = appCreator({
+    port: process.env.PORT,
+    loggingEnabled: process.env.WEBRTC_SIGNAL_LOGGING,
+    heartbeatEnabled: process.env.WEBRTC_HEARTBEAT_ENABLED,
```
Comment (anderm) [inline index.ts]: Switching to import will cause some errors here. Environment variables are always strings and this tries to cast it to a boolean. You can add a falsey check here, similar to the one in `lib/index.ts`

File: lib/index.ts (hunk 4 above)
Comment (conrad465) [reply lib/index.ts]: Is there a preference besides consistency?
removed the two semi colon lines

File: lib/index.ts (hunk 2 above)
Comment (conrad465) [reply lib/index.ts]: To import them we need an import of @Types/module which does not exist for these packages. I tried to make a universal import, but this error prevented it. I can manually add an index.d.ts file to node_modules, but I figured it was not a long term solution and as they were ported, we could switch the imports

File: dist/index.d.ts (hunk 1 above)
Comment (conrad465) [reply dist/index.d.ts]: Whoops! I added  dist to git ignore after this slipped through

File: lib/index.ts (hunk 2 above)
Comment (anderm) [reply lib/index.ts]: You only need to create a new file `modules.d.ts` at the root of the project, not inside node_modules. You then declare like this `declare module 'webrtc-signal-http-heartbeat';` You can do this for every import that you have. After you have the modules file, you can then import like this `import * as heartbeatRouterCreator  from "webrtc-signal-http-heartbeat";` and the error will go away. 

You might also need to add the `modules.d.ts` file in `tsconfig.json` under `include`

File: lib/index.ts (hunk 4 above)
Comment (anderm) [reply lib/index.ts]: My preference would be to add `;` on all lines, but this is more for consistency

File: lib/index.ts (hunk 2 above)
Comment (conrad465) [reply lib/index.ts]: done! easy fix. thank you!

File: lib/index.ts (hunk 4 above)
Comment (conrad465) [reply lib/index.ts]: done and added a linter!

File: lib/modules.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import { Express } from "express";
```
Comment (bengreenier) [inline lib/modules.ts]: nit: modules is a slightly confusing name, given https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/

File: modules.d.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+declare module 'webrtc-signal-http-heartbeat';
```
Comment (bengreenier) [inline modules.d.ts]: sorry, why is this needed?

File: tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    "compilerOptions": {
+      "module": "commonjs",
+      "declaration": true,
+      "target": "es6",
+      "noImplicitAny": true,
+      "sourceMap": true,
+      "outDir": "dist",
+      "baseUrl": ".",
+      "paths": {
+        "*": ["node_modules/*"]
```
Comment (bengreenier) [inline tsconfig.json]: my understanding is that this should be implicitly part of the lookup? is it not??

File: tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+      },
+      "resolveJsonModule": true
+    },
+    "include": [
+      "lib/**/*",
+      "index.ts",
```
Comment (bengreenier) [inline tsconfig.json]: preference to *.ts if possible

**Comment by anderm** [comment]
This looks good now!

File: lib/index.ts (hunk 2 above)
Comment (conrad465) [reply lib/index.ts]: @bengreenier including you here for context

File: modules.d.ts (hunk 13 above)
Comment (conrad465) [reply modules.d.ts]: Its a workaround for the fact that none of the webrtc-signal-http-foo deps are typescripted. I @-ed you in another comment for more context!

File: tsconfig.json (hunk 14 above)
Comment (conrad465) [reply tsconfig.json]: you are correct

File: lib/modules.ts (hunk 12 above)
Comment (conrad465) [reply lib/modules.ts]: renames to utils
